pkg/server/plugin/datastore/sql/sql.go

```diff
@@ -2694,65 +2689,91 @@ func applyPagination(p *datastore.Pagination, entryTx *gorm.DB) (*gorm.DB, error
 		return nil, sqlError.Wrap(err)
 	}
+	if req.Mask == nil || req.Mask.Selectors {
+		// Delete existing selectors - we will write new ones
+		if err := tx.Exec("DELETE FROM selectors WHERE registered_entry_id = ?", entry.ID).Error; err != nil {
+			return nil, sqlError.Wrap(err)
+		}
 
-	// Delete existing selectors - we will write new ones
-	if err := tx.Exec("DELETE FROM selectors WHERE registered_entry_id = ?", entry.ID).Error; err != nil {
-		return nil, sqlError.Wrap(err)
-	}
+		selectors := []Selector{}
+		for _, s := range req.Entry.Selectors {
+			selector := Selector{
+				Type:  s.Type,
+				Value: s.Value,
+			}
 
-	selectors := []Selector{}
-	for _, s := range req.Entry.Selectors {
-		selector := Selector{
-			Type:  s.Type,
-			Value: s.Value,
+			selectors = append(selectors, selector)
 		}
-
-		selectors = append(selectors, selector)
+		entry.Selectors = selectors
 	}
 
-	// Delete existing DNSs - we will write new ones
-	if err := tx.Exec("DELETE FROM dns_names WHERE registered_entry_id = ?", entry.ID).Error; err != nil {
-		return nil, sqlError.Wrap(err)
-	}
+	if req.Mask == nil || req.Mask.DnsNames {
+		// Delete existing DNSs - we will write new ones
+		if err := tx.Exec("DELETE FROM dns_names WHERE registered_entry_id = ?", entry.ID).Error; err != nil {
+			return nil, sqlError.Wrap(err)
+		}
 
-	dnsList := []DNSName{}
-	for _, d := range req.Entry.DnsNames {
-		dns := DNSName{
-			Value: d,
+		dnsList := []DNSName{}
+		for _, d := range req.Entry.DnsNames {
+			dns := DNSName{
+				Value: d,
+			}
+
+			dnsList = append(dnsList, dns)
 		}
+		entry.DNSList = dnsList
+	}
 
-		dnsList = append(dnsList, dns)
+	if req.Mask == nil || req.Mask.SpiffeId {
+		entry.SpiffeID = req.Entry.SpiffeId
+	}
+	if req.Mask == nil || req.Mask.ParentId {
+		entry.ParentID = req.Entry.ParentId
+	}
+	if req.Mask == nil || req.Mask.Ttl {
+		entry.TTL = req.Entry.Ttl
+	}
+	if req.Mask == nil || req.Mask.Admin {
+		entry.Admin = req.Entry.Admin
+	}
+	if req.Mask == nil || req.Mask.Downstream {
+		entry.Downstream = req.Entry.Downstream
+	}
+	if req.Mask == nil || req.Mask.EntryExpiry {
+		entry.Expiry = req.Entry.EntryExpiry
 	}
 
-	entry.SpiffeID = req.Entry.SpiffeId
-	entry.ParentID = req.Entry.ParentId
-	entry.TTL = req.Entry.Ttl
-	entry.Selectors = selectors
-	entry.Admin = req.Entry.Admin
-	entry.Downstream = req.Entry.Downstream
-	entry.Expiry = req.Entry.EntryExpiry
-	entry.DNSList = dnsList
 	if err := tx.Save(&entry).Error; err != nil {
 		return nil, sqlError.Wrap(err)
 	}
 
-	federatesWith, err := makeFederatesWith(tx, req.Entry.FederatesWith)
-	if err != nil {
-		return nil, err
+	if req.Mask == nil || req.Mask.FederatesWith {
+		federatesWith, err := makeFederatesWith(tx, req.Entry.FederatesWith)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := tx.Model(&entry).Association("FederatesWith").Replace(federatesWith).Error; err != nil {
+			return nil, err
+		}
```

Comment:
Do we need to set `entry.FederatesWith`? Otherwise it seems that the returned entry won't have the update?

Reply:
Hmm! You're right, this is kind of ugly... The call to modelToEntry will actually fill in the federatesWith field later. I will add a comment to indicate that.